google/cloud/bigtable/backup.py

```diff
@@ -0,0 +1,332 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A user-friendly container for Cloud Bigtable Backup."""
+
+import re
+
+from google.cloud._helpers import _datetime_to_pb_timestamp
+from google.cloud.exceptions import NotFound
+
+
+class Backup(object):
+    """Representation of a Google Cloud Bigtable Backup.
+
+    A :class: `Backup` can be used to:
+
+    * :meth:`create` the backup
+    * :meth:`update` the backup
+    * :meth:`delete` the backup
+
+    :type backup_id: str
+    :param backup_id: The ID of the backup.
+
+    :type instance: :class:`~google.cloud.spanner_v1.instance.Instance`
+    :param instance: The Instance that owns this Backup.
+
+    :type cluster_id: str
+    :param cluster_id: (Optional) The ID of the Cluster that contains this Backup.
+                       Required for calling 'delete', 'exists' etc. methods.
+
+    :type table_id: str
+    :param table_id: (Optional) The ID of the Table that the Backup is for.
+                     Required if the 'create' method will be called.
+
+    :type expire_time: :class:`datetime.datetime`
+    :param expire_time: (Optional) The expiration time after which the Backup
+                        will be automatically deleted. Required if the `create`
+                        method will be called.
+    """
+
+    def __init__(
+        self, backup_id, instance, cluster_id=None, table_id=None, expire_time=None
+    ):
+        self.backup_id = backup_id
+        self._instance = instance
+        self._cluster = cluster_id
+        self.table_id = table_id
+        self._expire_time = expire_time
+
+        self._parent = None
+        self._source_table = None
+        self._start_time = None
+        self._end_time = None
+        self._size_bytes = None
+        self._state = None
+
+    @property
+    def name(self):
+        """Backup name used in requests.
+
+        The Backup name is of the form
+
+            ``"projects/../instances/../clusters/../backups/{backup_id}"``
+
+        :rtype: str
+        :returns: The Backup name.
+
+        :raises: ValueError: If the 'cluster' has not been set.
+        """
+        if not self._cluster:
+            raise ValueError('"cluster" parameter must be set')
+
+        return "{}/clusters/{}/backups/{}".format(
```

Comment:
While it is true that using a 'canonical' API method instead of `format` is preferred, the `backup_path` only calls for `google.api_core.path_template.expand` method, while in order to raise a `ValueError`, the `google.api_core.path_template.validate` also needs to be called. This apparently does not happen when calling the `bigtable_table_admin_client.backup_path`, as confirmed via `TestBackup.test_restore_cluster_not_set`. So the ` if not self._cluster ...` clause must stay.